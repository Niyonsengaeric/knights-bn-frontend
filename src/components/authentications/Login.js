import React, { Component } from "react";
import { Button, Form, Image, Alert } from "react-bootstrap";
import "../../assets/styles/containers/loginPage.scss";
import { connect } from "react-redux";
import { fetchUsers } from '../../redux/actions/index';
import { Redirect } from 'react-router-dom';

// eslint-disable-next-line react/prefer-stateless-function
class Login extends Component {
    constructor(props) {
    super(props);
    this.state = { data: { email: '', password: '' } };
  }
  render() {
    return (
      <div className="login-page">
        <Form>
          <h1>Sign in</h1>
          <p>Login with so social media</p>
          <div className="Google-login-btn">
            {/* <Image src="../../assets/images/google.jpeg" rounded /> */}
            Sign in with Google
          </div>
          <div className="Facebook-login-btn">Sign in with Facebook</div>
          <div id="or">
		      	<svg className="Line" viewBox="0 1 120.5 1">
			        	<path fill="transparent" stroke="rgba(231,231,231,1)" width="1px" linejoin="miter" linecap="square" miterlimit="10" rendering="auto" id="Line" d="M 0 1 L 120.5 1">
			        	</path>
		      	</svg>
		      	<div id="or_">
			    	  <span>or</span>
		      	</div>
		      	<svg className="Line_" viewBox="0 1 120.5 1">
			    	  <path fill="transparent" stroke="rgba(231,231,231,1)" width="1px" linejoin="miter" linecap="square" miterlimit="10" rendering="auto" id="Line_" d="M 0 1 L 120.5 1">
			    	  </path>
		      	</svg>
		      </div>
          <Form.Group>
            <Form.Control
              type="email"
              placeholder="email"
              id="email"
              name="email"
              onChange={this.onChangeHandler}
            />
          </Form.Group>
          <Form.Group >
            <Form.Control
              type="password"
              placeholder="password"
              name="password"
              onChange={this.onChangeHandler}
              id="password"
            />
          </Form.Group>
          <p>{this.props.token ? <Redirect to="home" /> : null}</p>
          <p>{this.props.error}</p>
          <Button
            onClick={(e) => this.login(e)}
            variant="primary"
            size="lg"
            id="loginBtn"
            block
          >
            Login
          </Button>
          <a id="forgotPassword" href="#"><p>Forgot Password?</p></a>
        </Form>
      </div>
    );
  }
  onChangeHandler = (event) => {
    console.log(event);
    this.setState({
      data: {
        ...this.state.data,
        [event.target.name]: event.target.value,
      },
    });
  };
  login = (e) => {
    e.preventDefault();
    this.props.fetchUsers(this.state.data);
  };
}

const mapStateToProps = (state) => {
  return {
    token: state.user.token,
    error: state.user.message,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    fetchUsers: (request) => dispatch(fetchUsers(request)),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Login);
