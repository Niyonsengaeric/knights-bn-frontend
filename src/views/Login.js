import React, { Component } from 'react';
import { Form, Spinner } from 'react-bootstrap';
import '../assets/styles/containers/loginPage.scss';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import { withRouter } from 'react-router';
import { useTranslation } from 'react-i18next';
import jwt from 'jsonwebtoken';
import dotenv from 'dotenv';
import {
  validateLoginInput,
  loginUsersSuccess,
  handleError,
  setLoadingStatus,
} from '../redux/actions/index';
import TextBox from '../components/Textbox';
import Button from '../components/Button';
import Line from '../components/line';
import { callApiThunk as thunk } from '../redux/thunk/index';
import AlertComponent from '../components/Alert';
import SocialLogin from '../components/SocialLogin';
import 'react-universal-hooks';

dotenv.config();

export class Login extends Component {
  async handleClick(translate) {
    this.props.handleError('');
    const { validations: validatedFields } = this.props;
    const invalidFields = Object.keys(validatedFields).filter(
      (field) => validatedFields[field] === 'is-invalid',
    );
    const validatedFieldsToArray = Object.keys(validatedFields);
    if (validatedFieldsToArray.length === 2 && invalidFields.length === 0) {
      this.props.setLoadingStatus(true);
      await this.props.thunk(
        'post',
        '/auth/login',
        loginUsersSuccess,
        this.props.user,
      );
      this.props.setLoadingStatus(false);

      const { isLoggedIn, token } = this.props.userData;
      localStorage.setItem('user-token', token);

      if (isLoggedIn === false) return this.props.handleError('Incorrect email or password. Please try again.');
      const userRole = jwt.decode(token).role;
      const isSuperAdmin = (userRole === 'superAdmin');
      isLoggedIn && isSuperAdmin
        ? this.props.history.push('/admin')
        : this.props.history.push('/home');
    } else {
      const errorMsg = validatedFieldsToArray.length === 0
        ? translate('Please fill in the required fields.1')
        : this.props.validations.email === 'is-valid'
            && !this.props.validations.password
          ? translate('Please enter your password.1')
          : translate('Wrong email address.1');
      this.props.handleError(errorMsg);
    }
  }

  render() {
    const { t: translate } = useTranslation();
    const { validateInput, error, isLoading } = this.props;
    return (
      <div className="login-page loginContainer">
        <Spinner
          animation="border"
          variant="primary"
          className={isLoading ? 'spinner--position__center' : 'hide'}
        />
        <Form className="form">
          <h1>{translate('Sign in.1')}</h1>
          <AlertComponent isError={!!error} message={error} />
          <p>{translate('Login with social media.1')}</p>
          <SocialLogin
            googleAction={translate('Sign in with Google.1')}
            facebookAction={translate('Sign in with Facebook.1')}
          />
          <div id="or">
            <Line className="Line" />
            <div id="or_">
              <span>or</span>
            </div>
            <Line className="Line_" />
          </div>
          <TextBox
            type="email"
            placeholder={translate('email.1')}
            id="email"
            name="email"
            onChange={validateInput}
            label="email"
            value={this.props.user.email}
          />
          <TextBox
            type="password"
            placeholder={translate('password.1')}
            id="password"
            name="password"
            onChange={validateInput}
            label="password"
            value={this.props.user.password}
          />
          <p>{this.props.token ? <Redirect to="home" /> : null}</p>
          <Button
            aria-label="login"
            onClick={() => this.handleClick(translate)}
            id="loginBtn"
            label={translate('Login.1')}
          />
          <a id="forgotPassword" href="/forgetpassword">
            <p>{translate('Forgot Password.1')}</p>
          </a>
        </Form>
      </div>
    );
  }
}
export const mapStateToProps = (state) => ({
  token: state.user.token,
  error: state.errorHandler.error,
  validations: state.eventHandler.validations.login,
  user: state.eventHandler.user,
  userData: state.user.data,
  isLoading: state.eventHandler.isLoading,
});

const mapDispatchToProps = {
  validateInput: validateLoginInput,
  thunk,
  handleError,
  setLoadingStatus,
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Login));
