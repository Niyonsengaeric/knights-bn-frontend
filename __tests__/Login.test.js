import React from 'react';
import ReactDOM from 'react-dom';
import { cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';
import ReactTestRenderer from 'react-test-renderer';
import Login from '../src/components/Textbox';

describe('<Login />', () => {
  afterEach(cleanup);
  it('renders without crushing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Login />, div);
  });
  it('matches snapshot', () => {
    const tree = ReactTestRenderer.create(<Login />);
    expect(tree).toMatchSnapshot();
  });
});
