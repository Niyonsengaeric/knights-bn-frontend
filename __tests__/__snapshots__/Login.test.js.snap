// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`=============>LOGIN<============= should render login component 1`] = `
<DocumentFragment>
  <div
    class="login-page loginContainer"
  >
    <div
      class="hide spinner-border text-primary"
    />
    <form
      class=""
    >
      <h1>
        Sign in
      </h1>
      <span>
        <p
          class="hide"
          id="loginError"
        >
          <img
            class="alertIcon"
            src="[object Object]"
          />
        </p>
      </span>
      <p>
        Login with social media
      </p>
      <div
        class="Google-login-btn"
      >
        <img
          class="facebookLogo"
          src="[object Object]"
        />
        Sign in with Google
      </div>
      <div
        class="Facebook-login-btn"
      >
        <img
          class="GoogleLogo"
          src="[object Object]"
        />
        Sign in with Facebook
      </div>
      <div
        id="or"
      >
        <svg
          class="Line"
          viewBox="0 1 120.5 1"
        >
          <path
            d="M 0 1 L 120.5 1"
            fill="transparent"
            id="Line"
            linecap="square"
            linejoin="miter"
            miterlimit="10"
            rendering="auto"
            stroke="rgba(231,231,231,1)"
            width="1px"
          />
        </svg>
        <div
          id="or_"
        >
          <span>
            or
          </span>
        </div>
        <svg
          class="Line_"
          viewBox="0 1 120.5 1"
        >
          <path
            d="M 0 1 L 120.5 1"
            fill="transparent"
            id="Line"
            linecap="square"
            linejoin="miter"
            miterlimit="10"
            rendering="auto"
            stroke="rgba(231,231,231,1)"
            width="1px"
          />
        </svg>
      </div>
      <div
        class="form-group"
      >
        <input
          aria-label="email"
          class="form-control"
          id="email"
          name="email"
          placeholder="email"
          type="email"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <input
          aria-label="password"
          class="form-control"
          id="password"
          name="password"
          placeholder="password"
          type="password"
          value=""
        />
      </div>
      <p />
      <button
        aria-label="Login"
        class="btn btn-primary"
        id="loginBtn"
        label="Login"
        type="button"
      >
        Login
      </button>
      <a
        href="#"
        id="forgotPassword"
      >
        <p>
          Forgot Password?
        </p>
      </a>
    </form>
  </div>
</DocumentFragment>
`;
